src/herder/TransactionQueue.cpp

```diff
@@ -737,98 +743,116 @@ TransactionQueue::broadcastTx(AccountState& state, TimestampedTx& tx)
     state.mBroadcastQueueOps -= tx.mTx->getNumOperations();
     return mApp.getOverlayManager().broadcastMessage(
         tx.mTx->toStellarMessage());
 }
 
-bool
-TransactionQueue::broadcastSome()
+struct TxQueueTracker
 {
-    // broadcast transactions, sorted in apply-order per account to
-    // maximize chances of propagation.
-    size_t opsToFlood = getMaxOpsToFloodThisPeriod();
+    TransactionQueue::TimestampedTransactions::iterator mCur;
+    TransactionQueue::AccountState* mAccountState;
 
-    struct TxTracker
+    struct Comparator
     {
-        TimestampedTransactions::iterator mCur;
-        AccountState* mAccountState;
-        TxTracker(TxTracker&&) = default;
+        size_t mSeed;
+        Comparator(size_t seed) : mSeed(seed)
+        {
+        }
 
-        // NB: this overload seems to be necessary with most versions of
-        // libstdc++ that assert when std::move is called on self for iterators
-        // std::swap ends up calling this operator
-        // See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85828
-        TxTracker&
-        operator=(TxTracker&& other)
+        // return true if l < r
+        // compares tx `cur` for each TxTracker as to implement surge
+        // pricing comparison
+        bool
+        operator()(TxQueueTracker const& l, TxQueueTracker const& r) const
         {
-            if (this != &other)
+            if (l.mCur == l.mAccountState->mTransactions.end())
+            {
+                return r.mCur != r.mAccountState->mTransactions.end();
+            }
+            if (r.mCur == r.mAccountState->mTransactions.end())
             {
-                mCur = other.mCur;
-                mAccountState = other.mAccountState;
+                return false;
             }
-            return *this;
+            return lessThanXored(l.mCur->mTx, r.mCur->mTx, mSeed);
         }
     };
-    // populate the list of per account queues of transactions to broadcast
-    std::deque<TxTracker> iters;
+};
+
+bool
+TransactionQueue::broadcastSome()
+{
+    // broadcast transactions in surge pricing order:
+    // loop over transactions by picking from the account queue with the
+    // highgest base fee not broadcasted so far.
+    // This broadcasts from account queues in order as to maximize chances of
+    // propagation.
+    size_t opsToFlood = getMaxOpsToFloodThisPeriod();
+
+    // uses a priority queue of trackers, using a custom comparator as to
+    // prioritize higher fee queues
+    TxQueueTracker::Comparator comp(mBroadcastSeed);
+    std::priority_queue<TxQueueTracker, std::vector<TxQueueTracker>,
+                        TxQueueTracker::Comparator>
+        iters(comp);
     size_t totalOpsToFlood = 0;
     for (auto& m : mAccountStates)
     {
         auto asOps = m.second.mBroadcastQueueOps;
         if (asOps != 0)
         {
-            iters.emplace_back(
-                TxTracker{m.second.mTransactions.begin(), &m.second});
+            iters.emplace(
```

Comment:
This _almost_ produces the right broadcast order, but not exactly. There is one error. Example:
- `mAccountStates.size() == 2` so call the first element `m1` and the second `m2`
-  `m1` has 2 transactions each with 1 operation. The first has `fee = 100` and the second has `fee = 200` (the transaction queue limiter permits this as long as adding the higher fee wouldn't have required the lower fee to be removed). The first one has been broadcasted but the second one hasn't.
- `m2` has 1 transaction with 1 operation. It has `fee = 150` and has not been broadcasted (this is possible if it was added after the first `m1` transaction had already been broadcasted).
- The next transaction that should be broadcasted is the second `m1` transaction because it has the highest fee of all remaining transactions. But `m2` will be at the top of the priority queue because the `m2` transaction has a higher fee than the first `m1` transaction. Therefore the `m2` transaction will broadcast before the second `m1` transaction which is not in surge pricing order.

The `while` loop on 847,851 needs to be duplicated here. Note that the `while` loop still is needed at 847,851 to properly handle fee-bump replace-by-fee in the middle of a queue; otherwise the same situation described above can happen in that case. Please add a comment there to clarify why the loop is needed in both places. Furthermore, because the loop is being used twice I recommend that you factor it into a function.

Reply:
yes indeed great catch - a very subtle fairness issue